Enforce Pylint rules

DimitriPapadopoulos opened the pull request:
~~Also a couple Pylint rules that may not yet be implemented in ruff.~~

**Edit:** Run Pylint in addition to ruff, because not all Pylint rules have been or can be implemented in ruff.

TODO:
* [ ] Add unit tests and/or doctests in docstrings
* [ ] Add docstrings and API docs for any new/modified user-facing classes and functions
* [ ] New/modified features documented in `docs/user-guide/*.rst`
* [ ] Changes documented as a new file in `changes/`
* [x] GitHub Actions have all passed
* [x] Test coverage is 100% (Codecov passes)

dstansby commented:
I'm 👍 for spelling fixes, and 👎 for refactoring without a linter doing it automatically for us (because it just creates PR review noise for not much technical debt reduction). Could you open a PR to repalce this that just includes the spelling fixes?

DimitriPapadopoulos commented:
I think Sourcery applies linter rules, but rules that might not be part of ruff (yet). More importantly, I think Sourcery is a good addition - AI assistance for PR review. But I understand you don't agree with it.

Some of the Sourcery suggestions do improve source code, such as:
```diff
-    n_ellipsis = sum(1 for i in selection if i is Ellipsis)
+    n_ellipsis = selection.count(Ellipsis)
```
How about keeping the couple changes that really make sense?

DimitriPapadopoulos committed:
Apply ruff/refurb preview rule FURB140

FURB140 Use `itertools.starmap` instead of the generator

DimitriPapadopoulos committed:
Init variable in the branch where it's used

DimitriPapadopoulos committed:
Apply Pylint rule W2301 (unnecessary-ellipsis)

Unnecessary ellipsis constant

DimitriPapadopoulos committed:
Apply Pylint rule W0246 (useless-parent-delegation)

Useless parent or super() delegation in method '__init__'

DimitriPapadopoulos committed:
Apply Pylint rule W0404 (reimported)

DimitriPapadopoulos committed:
Apply Pylint rule C0325 (superfluous-parens)

Unnecessary parens after 'not' keyword
Unnecessary parens after ':=' keyword

DimitriPapadopoulos committed:
Apply Pylint rule R1708 (stop-iteration-return)

Do not raise StopIteration in generator, use return statement instead

DimitriPapadopoulos committed:
Add Pylint configuration

Disable rules that:
- we don't want to fix,
- have been implemenetd in ruff,
- could be enabled later and are currently followed by `FIXME`.

DimitriPapadopoulos committed:
CI: Enforce Pylint rules

DimitriPapadopoulos commented:
I have added Pylint to CI.

Adding it to pre-commit is discouraged because it is slow and incompatible with pre-commit parallelisation:
* [Pre-commit integration](https://pylint.pycqa.org/en/stable/user_guide/installation/pre-commit-integration.html)

DimitriPapadopoulos committed:
Apply Pylint rule W0201 (attribute-defined-outside-init)

Attribute '_is_open' defined outside __init__

DimitriPapadopoulos committed:
Apply Pylint rule W0108 (unnecessary-lambda)

Lambda may not be necessary

DimitriPapadopoulos committed:
Apply Pylint rule R1721 (unnecessary-comprehension)

Unnecessary use of a comprehension

DimitriPapadopoulos committed:
Apply Pylint rule R1716 (chained-comparison)

Simplify chained comparison between the operands

DimitriPapadopoulos committed:
Apply Pylint rule C0206 (consider-using-dict-items)

Consider iterating with .items()

DimitriPapadopoulos committed:
Apply Pylint rule W0231 (super-init-not-called)

__init__ method from base class is not called

DimitriPapadopoulos committed:
Apply Pylint rule C2801 (unnecessary-dunder-call)

Unnecessarily calls dunder method __init__. Instantiate class directly.
